content/blog/2021-02-01-e2e-verifier-design/index.md

```diff
@@ -0,0 +1,182 @@
+
+## Design Thinking
+
+We will implement the verifier with [Go template](https://golang.org/pkg/text/template/), plus some enhancements. Firstly, users need to write a Go template file with provided functions and actions to describe how the expected data looks like. Then the verifer renders the template with the actual data object. Finally, the verifier compares the rendered output with the actual data. If the rendered output is not the same with the actual output, it means the actual data is inconsist with the expected data. Otherwise, it means the actual data match the expected data. On failure, the verifier will also print out what are different between expected and actual data.
+
+## Branches / Actions
+
+The verifier inherits all the actions from the standard Go template, such as `if`, `with`, `range`, etc. In addition, we also provide some custom actions to satisfy our own needs.
+
+### At Least Once of List Elements Match
+
+`atLeastOnce` checks if there is at least one element of list matches the given template.
+
+Examples:
+
+```yaml
+metrics:
+{{- atLeastOnce .metrics }}
+  name: {{ notEmpty .name }}
+  id: {{ notEmpty .id }}
+  value: {{ gt .value .0 }}
+{{- end }}
+```
+
+Note that between the `{{- atLeastOnce .metrics }}` and `{{- end }}` it is a single object, instead of a list.
+
+## Functions
+
+Users can use these provided functions in the template to describe the expected data.
+
+### Not Empty
+
+`notEmpty` checks if the string `s` is empty.
+
+Example:
+
+```yaml
+id: {{ notEmpty .id }}
+```
+
+### Regexp match
+
+`regexp` checks if string `s` matches the regular expression pattern.
+
+Examples:
+
+```yaml
+label: {{ regexp .label "ratings.*" }}
+```
+
+### Base64
+
+`b64enc s` returns the Base64 encoded string of s.
+
+Examples:
+
+```yaml
+id: {{ b64enc "User" }}.static-suffix # this evalutes the base64 encoded string of "User", concatenated with a static suffix ".static-suffix"
+```
+
+Result:
+
+```yaml
+id: VXNlcg==.static-suffix
+```
+
+## Full Example
+
+Here is an example of expected data:
+
+```yaml
+# expected.data.yaml
+nodes:
+  - id: {{ b64enc "User" }}.0
+    name: User
+    type: USER
+    isReal: false
+  - id: {{ b64enc "Your_ApplicationName" }}.1
+    name: Your_ApplicationName
+    type: Tomcat
+    isReal: true
+  - id: {{ $h2ID := (index .nodes 2).id }}{{ notEmpty $h2ID }} # We assert that nodes[2].id is not empty and save it to variable `h2ID` for later use
+    name: localhost:-1
+    type: H2
+    isReal: false
+calls:
+  - id: {{ notEmpty (index .calls 0).id }}
+    source: {{ b64enc "Your_ApplicationName" }}.1
+    target: {{ $h2ID }} # We use the previously assigned variable `h2Id` to asert that the `target` is equal to the `id` of the nodes[2]
+    detectPoints:
+      - CLIENT
+  - id: {{ b64enc "User" }}.0-{{ b64enc "Your_ApplicationName" }}.1
+    source: {{ b64enc "User" }}.0
+    target: {{ b64enc "Your_ApplicationName" }}.1
+    detectPoints:
+      - SERVER
+```
+
+will validate this data:
```

Comment:
I want to be clear one more thing, does this verification will compare the size of the list? The logic behind the question also includes, do we check every rule has one and only one matched data. Or just at least having one matched.

Reply:
> I want to be clear one more thing, does this verification will compare the size of the list? The logic behind the question also includes, do we check every rule has one and only one matched data. Or just at least having one matched.

Rules inside `{{ atLeastOnce }}` blocks will do an "at least once" verification, otherwise they are compared / verified literally (include list size of course).

This reminds me that we need another action `{{ ignoreOrders }}` that can verify 2 lists ignoring their elements' orders @fgksgf , please add this to the doc.

I think this covers the general use cases of lists comparison.

Reply:
> Rules inside {{ atLeastOnce }} blocks will do an "at least once" verification, otherwise they are compared / verified literally (include list size of course).

`atLeastOnce` so this `at least` is from the rule's perspective? rather than data at least matched one rule?

Reply:
In some cases, you could have 4 rules to match 5+ data, if you use wildcards. Is this expected to pass? And, is 5 rules to match 4- list, but all 4- elements are checked at least by one rule, is this expected to pass?
I think we need to choose the word carefully, and be clear about the logic in this doc.

Reply:
> > Rules inside {{ atLeastOnce }} blocks will do an "at least once" verification, otherwise they are compared / verified literally (include list size of course).
> 
> `atLeastOnce` so this `at least` is from the rule's perspective? rather than data at least matched one rule?

Exactly, this is a typical case in our existing E2E (Java version). The latter one you said 👆 is more like a logical operation `OR` between rules, right? The name may be ambiguous indeed, let's rename it to `contains`, so the rule file may be something like

```yaml
metrics:
{{ contains .metrics }}
  name: {{ notEmpty .name }}
  value: {{ gt .value 0 }}
{{ end }}
```

 make sense? @wu-sheng , @fgksgf

Reply:
> In some cases, you could have 4 rules to match 5+ data, if you use wildcards. Is this expected to pass? And, is 5 rules to match 4- list, but all 4- elements are checked at least by one rule, is this expected to pass?
> I think we need to choose the word carefully, and be clear about the logic in this doc.

IMHO, all these cases can be essentially resolved with the support of `AND` / `OR` combinations of the basic rules, right? We can add that in a later version when we really need them (or community asks for that). Let me know if this resolves your concern.

Reply:
> when we really need them

I don't think we need them, I just hope it is clear.

Reply:
@fgksgf please polish the doc and clarify this part, ping me if you have any question

Reply:
ok